vapi/appliance/simulator/simulator.go

```diff
@@ -0,0 +1,191 @@
+	URL        *url.URL
+	consolecli struct {
+		access consolecli.Access
+		sync.RWMutex
+	}
+	dcui struct {
+		access dcui.Access
+		sync.RWMutex
+	}
+	ssh struct {
+		access ssh.Access
+		sync.RWMutex
+	}
+	shell struct {
+		access shell.Access
+		sync.RWMutex
+	}
+}
+
+// New creates a Handler instance
+func New(u *url.URL) *Handler {
+	return &Handler{
+		URL: nil,
+		consolecli: struct {
+			access consolecli.Access
+			sync.RWMutex
+		}{},
+		dcui: struct {
+			access dcui.Access
+			sync.RWMutex
+		}{},
+		ssh: struct {
+			access ssh.Access
+			sync.RWMutex
+		}{},
+		shell: struct {
+			access shell.Access
+			sync.RWMutex
+		}{},
+	}
+}
+
+// Register Appliance Management API paths with the vapi simulator's http.ServeMux
+func (h *Handler) Register(s *simulator.Service, r *simulator.Registry) {
+	s.HandleFunc(consolecli.Path, h.consoleCLIStatus)
+	s.HandleFunc(dcui.Path, h.dcuiStatus)
+	s.HandleFunc(ssh.Path, h.sshStatus)
+	s.HandleFunc(shell.Path, h.shellStatus)
+}
+
+func (h *Handler) decode(r *http.Request, w http.ResponseWriter, val interface{}) bool {
+	return Decode(r, w, val)
+}
+
+// Decode the request Body into val.
+// Returns true on success, otherwise false and sends the http.StatusInternalServerError response.
+func Decode(request *http.Request, writer http.ResponseWriter, val interface{}) bool {
+	defer request.Body.Close()
+	err := json.NewDecoder(request.Body).Decode(val)
+	if err != nil {
+		log.Printf("%s %s: %s", request.Method, request.RequestURI, err)
+		writer.WriteHeader(http.StatusInternalServerError)
+		return false
+	}
+	return true
+}
+
+func (h *Handler) consoleCLIStatus(writer http.ResponseWriter, request *http.Request) {
+	switch request.Method {
+	case http.MethodGet:
+		h.consolecli.RWMutex.RLock()
+		vapi.StatusOK(writer, h.consolecli.access.Enabled)
+		h.consolecli.RWMutex.RUnlock()
+	case http.MethodPut:
+		var input consolecli.Access
+		if h.decode(request, writer, &input) {
+			h.consolecli.RWMutex.Lock()
+			h.consolecli.access.Enabled = input.Enabled
+			h.consolecli.RWMutex.Unlock()
+			writer.WriteHeader(http.StatusNoContent)
+		}
+	default:
+		http.NotFound(writer, request)
+	}
+}
+
+func (h *Handler) dcuiStatus(writer http.ResponseWriter, request *http.Request) {
+	switch request.Method {
+	case http.MethodGet:
+		h.dcui.RLock()
+		vapi.StatusOK(writer, h.dcui.access.Enabled)
+		h.dcui.RUnlock()
+	case http.MethodPut:
+		var input dcui.Access
+		if h.decode(request, writer, &input) {
+			h.dcui.Lock()
+			h.dcui.access.Enabled = input.Enabled
+			h.dcui.Unlock()
+			writer.WriteHeader(http.StatusNoContent)
+		}
+	default:
+		http.NotFound(writer, request)
+	}
+}
+
+func (h *Handler) sshStatus(writer http.ResponseWriter, request *http.Request) {
+	switch request.Method {
+	case http.MethodGet:
+		h.ssh.RLock()
+		vapi.StatusOK(writer, h.ssh.access.Enabled)
+		h.ssh.Unlock()
+	case http.MethodPut:
+		var input ssh.Access
+		if h.decode(request, writer, &input) {
+			h.ssh.Lock()
```

Comment:
Either RLock or Unlock - but this combination panics

Reply:
I wonder why this did not panic in your tests? Is this covered?

Reply:
Tests are covered and and test had passed locally
```
ok 174 vcsa.access.consolecli.get
ok 175 vcsa.access.consolecli.set
ok 176 vcsa.access.dcui.get
ok 177 vcsa.access.dcui.set
ok 178 vcsa.access.shell.get
ok 179 vcsa.access.shell.set
ok 180 vcsa.access.ssh.get
ok 181 vcsa.access.ssh.set
```

Reply:
IMHO it's still a bug: https://go.dev/play/p/TtnI1BrYxfv so your test seems to not trigger it

Reply:
I think this might be due to your `$PATH` since you already have `govc` and `vcsim` in your path likely which are picked up instead. The `Makefile` does build both.

Reply:
Do you think, when we run `make govc-test` we should build vcsim and govc and put them in `govmomi/bin/` and run the govc-test pointing to these binaries?

Reply:
can we make tests to pick the binaries from the bin directory instead of $PATH?

Reply:
IMHO CI is fine because both binaries are not installed there. Locally for tests you can tweak `$PATH` before executing? If that helps, perhaps we just need to amend the CONTRIBUTIN guide?

Reply:
Sure, let me amend CONTRIBUTING guide

Reply:
Created a new issue to track https://github.com/vmware/govmomi/issues/2752